### PR TITLE
Bugbash

### PR DESCRIPTION
**Pull request by ARMmaster17**
This PR contains the following changes:
- Migrated DB operations to backend. #63 
- Split CI tests across 3 VMs.
- Localized dates and times in views. #55 
- Fixed certain pages treating `author` as a locale in URI. #52 
- Moved notification logic to API and frontent JS code.
- Added UglifyJS as asset compilation step.
- Added data caching with wrappers included in the `/lib` folder.
- Added memcacher/redis to development environment setup script.
- Added job tracking with Redis-Status.
- Better module/class method documentation.

**Merging this commit will finalize the 1.0.0 milestone!**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,38 +1,63 @@
 language: ruby
-cache: bundler
 rvm:
 - 2.4.0
 git:
   submodules: false
 addons:
   code_climate:
     repo_token: 3a37b06b904d7cd224d38b486334dd1c77c6eedf5efc96173531a470e9b8c8ea
+cache:
```
Comment (ARMmaster17) [inline .travis.yml]: Need to cut this out and revert back to bundler caching.

File: .travis.yml [hunk 2]
```diff
@@ -1,38 +1,63 @@
...
 env:
   global:
   - RACK_ENV=test
+  - IP=0.0.0.0
   - PORT=8080
   - APP_SUPER_UNAME=testuser
   - APP_SUPER_PASSWD=testpassword
+  - REDIS_URL=redis://localhost:6379
+  - MEMCACHIER_SERVERS=localhost:11211
+  - PHANTOMJS_VERSION=2.1.14
```
Comment (ARMmaster17) [inline .travis.yml]: Can remove obsolete variable.

File: Gemfile [hunk 3]
```diff
@@ -14,27 +14,35 @@ gem 'sinatra-r18n'
...
 gem 'rspec'
 gem 'capybara'
 gem 'poltergeist'
+
 group :development do
     gem 'guard'
     gem 'guard-migrate'
     gem 'guard-sass'
     gem 'guard-rubocop'
+    gem 'foreman'
```
Comment (ARMmaster17) [inline Gemfile]: Foreman needs to be in the development/testing group.

File: jobs/model_create.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    module Models
+        class Create
+            include Resque::Plugins::Status
+            
+            #@queue = :models
+            
+            def perform
+                # Find object in database.
+                begin
+                    @type = options['model_type'].split('::').reduce(Module, :const_get)
```
Comment (ARMmaster17) [inline jobs/model_create.rb]: Should document what is going on here since this is dynamic programming which is hard to follow outside context.

File: jobs/model_create.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+require_relative '../models/articles'
+require_relative '../models/feeds'
+require_relative '../models/resources'
+
+module Jobs
+    module Models
+        class Create
+            include Resque::Plugins::Status
+            
+            #@queue = :models
```
Comment (ARMmaster17) [inline jobs/model_create.rb]: May want to remove obsolete code, or figure out how to assign a Resque-Status job to a queue.

File: jobs/model_create.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            def perform
+                # Find object in database.
+                begin
+                    @type = options['model_type'].split('::').reduce(Module, :const_get)
+                    at(1, 3)
+                    @type.create!(options['args'])
+                    at(2, 3)
+                rescue ActiveRecord::RecordNotFound
+                    # Invalid ID.
+                    failed(msg: 'Model not found')
```
Comment (ARMmaster17) [inline jobs/model_create.rb]: Now that the frontend feeds directly from the job status, need to send the translated string. (Locale code will need to be passed through GET parameter.)

File: jobs/model_edit.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                if options['user_id'] == @obj.author || options['user_id'] == 'super' || options['is_super']
+                    # Update progress.
+                    at(2,4)
+                    # Update the object.
+                    begin
+                        @obj.update!(options['args'])
+                    rescue ActiveRecord::RecordInvalid
+                        failed(msg: 'Invalid parameters')
+                        return
+                    rescue
```
Comment (ARMmaster17) [inline jobs/model_edit.rb]: Shouldn't be suppressing exceptions with generic catch statement.

File: jobs/reload_cache.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    module Cache
+        class ReloadCache
+            
+            @queue = :cache
+            
+            def self.perform()
+                # Enqueue all cache reload subjobs.
+                Resque.enqueue(ReloadCacheHome)
+                # Update the timestamp on the cache.
+                Lib::Cache.set("cache-all-updated", Time.now.getutc)
```
Comment (ARMmaster17) [inline jobs/reload_cache.rb]: Should either actually utilize this function or remove the timestamps entirely and do event-based updates.

File: lib/cache.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+require 'dalli'
+require 'resque'
+require_relative '../jobs/reload_cache'
+
+module Lib
+    module Cache
+        
+        # Create a static Dalli object to handle Memcache connections.
+        DC = Dalli::Client.new(ENV['MEMCACHIER_SERVERS'], { :namespace => "wulfusa", :compress => true })
```
Comment (ARMmaster17) [inline lib/cache.rb]: Memcachier namespace should not be hardcoded.

File: models/articles.rb [hunk 10]
```diff
@@ -5,4 +5,15 @@ class Article < ActiveRecord::Base
         presence: true
     validates :content,
         presence: true
+    
+    ##
+    # Generates cache hash of object.
+    def generate_metadata
+        data = Hash.new
```
Comment (ARMmaster17) [inline models/articles.rb:12]: Can probably be compressed with an ActiveRecord helper.

File: public/src/js/jobtracking.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+var taskID = setInterval(function() {
+    var div = document.getElementById('msgHolder'),
+        divChildren = div.childNodes;
+    if (divChildren.length === 0) {
+        clearInterval(taskID);
+    }
+    for (var i = 0; i < divChildren.length; i++) {
+        $.get(
+            "/api/v1/job/" + divChildren[i].getAttribute('value'), {},
+            function(data) {
```
Comment (ARMmaster17) [inline public/src/js/jobtracking.js:14]: CC says this is a bad idea by putting a function in a loop. In reality, there's a 50% chance this code will break if there is more than one notification to present to the user (currently not possible).

File: routes/videos.rb [hunk 12]
```diff
@@ -12,6 +12,7 @@ def self.registered(app)
               ##
               # Blog listing of site.
               app.get '/:locale/videos' do
+                if params[:locale] == "author" then redirect "/#{locale?}/author/videos" end
```
Comment (ARMmaster17) [inline routes/videos.rb]: I don't know why this works, but it does so I'm leaving it.

File: spec/spec_helper.rb [hunk 13]
```diff
@@ -1,11 +1,13 @@
 require 'capybara/dsl'
 require 'capybara/poltergeist'
+require 'rspec-benchmark'
 require 'simplecov'
 SimpleCov.start
 require_relative '../wulf_app.rb'
 
 Capybara.default_driver = :poltergeist
-Capybara.app = proc { |env| WulfApp.new.call(env) }
+Capybara.app_host = 'http://localhost:8080'
```
Comment (ARMmaster17) [inline spec/spec_helper.rb]: App is isolated from RSpec thread, which may hurt code coverage. Should have it set up so that everthing is inline, unless the CI environment is testing performance.

File: spec/unit_cva_spec.rb [hunk 14]
```diff
@@ -1,12 +1,13 @@
...
-describe WulfApp do
+describe WulfApp, :unit do
     it "CVA-001: CSRF in SSO authentication" do
         visit "/en/sso/author/login"
         page.driver.clear_cookies
         fill_in 'Username', with: 'testuser'
         fill_in 'Password', with: 'testpassword'
         click_on 'Sign in'
         expect(page).to have_current_path "/en/sso/author/login"
+        #expect(page).to have_content("You have been subjected to XSRF attack. Contact the site administrators immediately.")
```
Comment (ARMmaster17) [inline spec/unit_cva_spec.rb:11]: This line doesn't work for some reason, but the intended behavior is still being tested so this can be ignored.

File: views/layout.slim [hunk 15]
```diff
@@ -38,12 +38,18 @@
       <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
     <![endif]-->
     <link rel="stylesheet" href="/src/css/site.css">
+    <script src="/src/js/jobtracking.min.js"></script>
```
Comment (ARMmaster17) [inline views/layout.slim:41]: Should probably have some kind of `<noscript>` tag for users that have JavaScript disabled.

**Comment by ARMmaster17** [comment]
Hold being placed until immediate issues are fixed.

**Comment by ARMmaster17** [comment]
List of changes requested by @ARMmaster17 (16):
- [x] Add Bundler caching to Travis config file.
- [x] Remove PhantomJS stuff from Travis config file.
- [x] Move foreman to dev/test group in Gemfile.
- [x] Document dynamic programming in model jobs files.
- [x] Assign model jobs to a queue.
- [x] Use translated strings for job outputs.
- [x] Make sure no exceptions are being suppressed.
- [x] Utilize or remove cache timestamp.
- [x] Move memcacher namespace to environment variable.
- [x] See if certain jobs can be compressed and secured using a Rails helper.
- [x] Remove function from loop in jobstatus JavaScript file.
- [x] Verify `author` locale bug patch is working as intended.
- [x] Verify notifications are getting passed with above bug patch.
- [x] Make tests work inline unless it is a performance test.
- [x] Figure out why completion message isn't being shown in testing environment.
- [x] Have noscript tag on pages for non-standard browser configurations.

File: lib/cache.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
+require 'dalli'
+require 'resque'
+require_relative '../jobs/reload_cache'
+
+module Lib
+    module Cache
+        
+        # Create a static Dalli object to handle Memcache connections.
+        DC = Dalli::Client.new(ENV['MEMCACHIER_SERVERS'], { :namespace => ENV['MEMCACHIER_NAMESPACE'], :compress => true })
```
Comment (ARMmaster17) [inline lib/cache.rb:9]: The Dalli documentation claims that we don't need to pass in a username and a password if we are on Heroku. So if something breaks in the future with caching, start looking here.

**Review by wulf502 — APPROVED** [review]
